Use higher order MappingQ in MappingQEulerian

gassmoeller opened the pull request:
This is a WIP pull request for @tiannh7 to test.

Do not review this yet, I need to clean up the code and add my test. The essence of the PR is that I think the MappingQEulerian class promises to work for higher order mappings, but I does not actually use a higher order mapping when computing the deformed mapping support points. This leads to poor accuracy e.g. if the underlying manifold is curved.

tiannh7 commented:
Hi Gassmoeller
Thank you for sharing this WIP pull request. I appreciate you preparing this update for me to test and look forward to testing the cleaned-up version once it is ready:).

marcfehling commented:
This would also be a useful addition to the [`MappingFEField`](https://www.dealii.org/developer/doxygen/deal.II/classMappingFEField.html) class.

tjhei reviewed (approved):
Makes sense!

kronbichler commented:
/rebuild

gassmoeller commented:
Ok, I cleaned up the code and added a test case. I also removed some comments in the source file that didnt seem to follow our usual convention of comments in source files.

This is ready from my side unless there are further requests.

@marcfehling I had a look at `MappingFEField`, but I didnt find a place where it uses FEValues. I think it may not apply to `MappingFEField`, because it expects the finite element field to provide all actual positions of the support points, so if the FE field is of higher order it should contain all the necessary information. The problem in `MappingQEulerian` was that it was using a first order base geometry and a (potentially higher order) deformation vector. But then I dont know `MappingFEField` very well, maybe I just didnt see the correct place that needed to be modified. If you point me in the right direction I can make a follow up PR for `MappingFEField`.

marcfehling commented:
@gassmoeller I was checking the constructors of the `MappingFEField` class, which instantiate `FEValues` objects for example here:
https://github.com/dealii/dealii/blob/efd041e05ee7669aa80183be36ac4f8c872a7951/source/fe/mapping_fe_field.cc#L290-L292

However, `MappingFEField` only uses [nodal_type_quadrature](https://www.dealii.org/developer/doxygen/deal.II/classReferenceCell.html#a61f5e71c69c2cc9f0a015ce4c434fa09). I don't think a higher order mapping is in any way beneficial in this case. So I would like to retract my statement :grimacing:

marcfehling commented:
@gassmoeller Maybe it would make sense to use [`MappingFE`](https://www.dealii.org/developer/doxygen/deal.II/classMappingFE.html) with the finite element from `this->euler_dof_handler->get_fe()` for the `FEValues` objects in the `MappingFEField` class.

drwells commented:
I know less about this class than `MappingFEField` but the behavior it has (of ignoring manifold information) is what I want it to do: i.e., when using that class I expect the finite element field to completely describe the geometry.

gassmoeller commented:
@drwells do you mean `MappingFEField` or `MappingQEulerian`? I couldnt quite figure out from your comment if you said I broke your use case, or if you meant we shouldnt modify `MappingFEField` because you use it without manifold information.
I am ok with keeping `MappingFEField` the way it is at the moment. I only need the modification for `MappingQEulerian`, and I think for this class using the manifold makes sense, because you would expect that `MappingQEulerian` behaves exactly as `MappingQ` if the deformation vector is 0. So do you agree we can keep the current state without further modification to other classes?
